app/models/users/scopes/having_reminder_mail_to_send.rb

```diff
@@ -0,0 +1,132 @@
+      #   This is done because time zones always have a mod(15) == 0 minutes offset.
+      #   Needs to be before the current time.
+      def having_reminder_mail_to_send(earliest_time)
+        local_times = local_times_from(earliest_time)
+
+        return none if local_times.empty?
+
+        # Left outer join as not all user instances have preferences associated
+        # but we still want to select them.
+        recipient_candidates = User
+                                 .active
+                                 .left_joins(:preference)
+                                 .joins(local_time_join(local_times))
+
+        subscriber_ids = Notification
+                           .unsent_reminders_before(recipient: recipient_candidates, time: Time.current)
+                           .group(:recipient_id)
+                           .select(:recipient_id)
+
+        where(id: subscriber_ids)
+      end
+
+      def local_time_join(local_times)
+        # Joins the times local to the user preferences and then checks whether:
+        # * reminders are enabled
+        # * any of the configured reminder time is the local time
+        # If no time zone is present, utc is assumed.
+        # If no reminder settings are present, sending a reminder at 08:00 local time is assumed.
+        <<~SQL.squish
+          JOIN (
+           SELECT * FROM #{arel_table.grouping(Arel::Nodes::ValuesList.new(local_times)).as('t(time, zone)').to_sql}
+          ) AS local_times
+          ON COALESCE(user_preferences.settings->>'time_zone', 'UTC') = local_times.zone
+          AND (
+            (
+              user_preferences.settings->'daily_reminders'->'times' IS NULL
+              AND local_times.time = '08:00:00+00:00'
+            )
+            OR
+            (
+              (user_preferences.settings->'daily_reminders'->'enabled')::boolean
+              AND user_preferences.settings->'daily_reminders'->'times' ? local_times.time
+            )
+          )
+        SQL
+      end
+
+      def local_times_from(earliest_time)
+        times = quarters_between_earliest_and_now(earliest_time)
+
+        times_for_zones(times)
+      end
+
+      def times_for_zones(times)
+        ActiveSupport::TimeZone
+          .all
+          .map do |z|
+            times.map do |time|
+              local_time = time.in_time_zone(z)
+
+              # Since only full hours can be configured, we can disregard any local time that is not
+              # a full hour.
+              next if local_time.min != 0
+
+              [local_time.strftime('%H:00:00+00:00'), z.name.gsub("'", "''")]
+            end
+          end
+          .flatten(1)
+          .compact
+      end
+
+      def quarters_between_earliest_and_now(earliest_time)
+        latest_time = Time.current
+        raise ArgumentError if latest_time < earliest_time || (latest_time - earliest_time) > 1.day
```

Comment:
What happens with jobs that get an `ArgumentError`? Will they simply drop from the jobs list (would be good here)? Or will they get retried (would be bad here).

Reply:
Yes, they will be dropped, since we only execute jobs once. In any case, I don't expect this to happen at all because that would mean that delayed job is broken. The only scenario I can currently envision is when the job is executed by hand preemptively e.g. when debugging.